drivers/mspi/mspi_stm32_xspi.c

```diff
@@ -0,0 +1,1903 @@
+			break;
+		default:
+			/* Not correct */
+			hal_ret = HAL_BUSY;
+			break;
+		}
+	}
+
+	if (hal_ret != HAL_OK) {
+		goto end;
+	}
+
+	/* Lock again expecting the IRQ for end of Tx or Rx */
+	if (k_sem_take(&dev_data->sync, K_FOREVER) < 0) {
+		LOG_ERR("Failed to take sem");
+		return HAL_BUSY;
+	}
+
+end:
+	if (hal_ret != HAL_OK) {
+		LOG_ERR("Failed to access data");
+	}
+
+	return hal_ret;
+}
+/**
+ * @brief Sends a Command to the NOR and Receive/Transceive data if relevant in IT or DMA mode.
+ *
+ */
+static int mspi_stm32_xspi_access(const struct device *dev, struct mspi_xfer_packet *packet,
+				  uint8_t access_mode)
+{
+	HAL_StatusTypeDef hal_ret;
+	struct mspi_stm32_data *dev_data = dev->data;
+
+	if (dev_data->xip_cfg.enable) {
+		if ((packet->cmd == SPI_NOR_CMD_WREN) || (packet->cmd == SPI_NOR_OCMD_WREN) ||
+		    (packet->cmd == SPI_NOR_CMD_SE_4B) || (packet->cmd == SPI_NOR_OCMD_SE) ||
+		    (packet->cmd == SPI_NOR_CMD_SE) ||
+		    ((mspi_stm32_xspi_hal_address_size(dev_data->dev_cfg.addr_length) ==
+		      HAL_XSPI_ADDRESS_24_BITS) &&
+		     (dev_data->dev_cfg.io_mode == MSPI_IO_MODE_SINGLE))) {
+			LOG_DBG(" MSPI_IO_MODE_SINGLE in 3Bytes addressing is not supported in "
+				"memory map mode, switching to indirect mode");
+
+			int ret = mspi_stm32_xspi_abort_memmap_if_enabled(dev);
+
+			if (ret != 0) {
+				return ret;
+			}
+			goto indirect;
+		}
+
+		if (read_write_in_memory_map_mode(dev, packet) == -EPROTONOSUPPORT) {
+			goto indirect;
+		}
+		return 0;
+	}
+
+indirect:
+	(void)pm_device_runtime_get(dev);
+	/* Prevent the clocks to be stopped during the request */
+	pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+
+	XSPI_RegularCmdTypeDef cmd =
+		mspi_stm32_xspi_prepare_cmd(dev_data->dev_cfg.io_mode, dev_data->dev_cfg.data_rate);
+
+	cmd.DataLength = packet->num_bytes;
+	cmd.Instruction = packet->cmd;
+	if (packet->dir == MSPI_TX) {
+		cmd.DummyCycles = dev_data->ctx.xfer.tx_dummy;
+	} else {
+		cmd.DummyCycles = dev_data->ctx.xfer.rx_dummy;
+	}
+	cmd.Address = packet->address; /* AddressSize is 32bits in OPSI mode */
+	cmd.AddressWidth = mspi_stm32_xspi_hal_address_size(dev_data->ctx.xfer.addr_length);
+	if (cmd.DataLength == 0) {
+		cmd.DataMode = HAL_XSPI_DATA_NONE;
+	}
+
+	if ((cmd.Instruction == SPI_NOR_CMD_WREN) || (cmd.Instruction == SPI_NOR_OCMD_WREN)) {
+		/* Write Enable only accepts HAL_XSPI_ADDRESS_NONE */
+		cmd.AddressMode = HAL_XSPI_ADDRESS_NONE;
+	}
+
+	LOG_DBG("MSPI access Instruction 0x%x", cmd.Instruction);
+
+	hal_ret = HAL_XSPI_Command(&dev_data->hmspi.xspi, &cmd, HAL_XSPI_TIMEOUT_DEFAULT_VALUE);
+	if ((hal_ret != HAL_OK) || (packet->num_bytes == 0)) {
+		pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+		(void)pm_device_runtime_put(dev);
+		if (hal_ret != HAL_OK) {
+			LOG_ERR("Failed to send XSPI instruction");
+			return -EIO;
+		}
+		if (packet->num_bytes == 0) {
+			return 0;
+		}
```

Comment:
To prevent confusion, I would suggest to use here either `else { return 0; }` or straight `return 0;`.